### ISSUE TITLE
Support multiple prefixes removal

### ISSUE THREAD

**Issue by andzej505hq** [jenkinsci/aws-secrets-manager-credentials-provider-plugin#227]
### What feature do you want to see added?

The current implementation of `transformations.name.removePrefix` only supports a single prefix to be removed. This works as I would expect, but it is limiting. I have multiple "groups" of secrets in a single AWS account, i.e. `jenkins/generic/<secret_name>`, `jenkins/test/<secret_name>`. I would like to remove both prefixes, but I don't see a way to do so...
Any suggestions for how to work around this will be appreciated

**Comment by andzej505hq** [comment]
I went spelunking in the code and found that the prefix is removed using a regex pattern compiled with `literal` flag. Does anything prevent you from using regex without that flag there?

**Comment by chriskilding** [comment]
The section in question is

```
if (str.startsWith(canonicalPrefix)) {
            return Pattern.compile(canonicalPrefix, Pattern.LITERAL)
                    .matcher(str)
                    .replaceFirst("");
} else {
            return str;
}
```

Off the top of my head I can't think why Pattern.literal is used in particular.

The bigger thing is that this regex match is nested inside the outer guard, `if string.startsWith(<prefix>)`, which is what determines overall behaviour here. Therefore setting a regex pattern as the prefix would not currently work.

I did this initially because (a) the use case at the time required removing a single prefix, and (b) I didn't want to invite regex into the system unless necessary because it complicates thing.

But things change and I'm potentially open to allowing regex patterns to be set, if this is a thing people need.

**Comment by andzej505hq** [comment]
I understand why you were hesitant to add regex in. It may be overkill for most cases and it often adds more problems than it fixes.
Would you be open to making the `transformations.name.removePrefix` node accept a list of prefixes? It would satisfy my use-case 100%

**Comment by chriskilding** [comment]
Yep, multiple prefixes in a list would be a reasonable way of doing it. Let's rustle up a PR...

**Comment by chriskilding** [comment]
Here's a couple of options for the design...

## Allow the `removePrefix` strategy to accept regexes

This is the smallest change. However it may be unintuitive to read, because even if the user does not specify `/^` at the start of the regex, it's always treated as a prefix match.

```yaml
awsCredentialsProvider:
  transformations:
    name:
      removePrefix:
        prefix: /(staging|production)-/
```

## The no-regex option

This is the one we discussed. Take multiple prefixes, don't accept regex patterns.

```yaml
awsCredentialsProvider:
  transformations:
    name:
      removePrefix:
        prefixes:
        - "staging-"
        - "production-"
```

It would be possible to write a Jenkins config migrator which will migrate the single prefix currently stored in the `prefix` field into the new format, which will help manually-managed Jenkins installations (but won't help CasC-managed installations):

```
this.prefixes = Arrays.asList(this.prefix);
this.prefix = null;
save();
```

## The full regex option

We create a new, more generically named `remove` strategy which accepts a `pattern`. The pattern could match the start of the secret name (as shown), or any part of the name, depending on what the user wants.

```yaml
awsCredentialsProvider:
  transformations:
    name:
      remove:
        pattern: /^(staging|production)-/
```

It *might* be possible to write a config migrator which would migrate existing `removePrefix` instances to this new format, and then deprecate the `removePrefix` strategy, but I'm not sure.

---

I have previously indicated a preference for the no-regex option, but it's possible that denying people the ability to use regex when they need it is just obtuse, rather than helpful. The old design adage that something should be 'as simple as possible, but no simpler' comes to mind.

**Comment by chriskilding** [comment]
Any thoughts on the above?

**Comment by andzej505hq** [comment]
I personally prefer specifying a list of prefixes to remove. It makes it easy to read, understand _and_ it has little room for error. Can say from experience that regexes can be easily misconfigured to match too much or too little.

Some can argue, that in an ideal case we would have both - ability to specify a regex and an ability to specify a list of prefixes. IMO having just a list would cover most of the cases, as most often we are working with pre-defined groups.

**Comment by chriskilding** [comment]
Fortunately the name transformer interface is polymorphic, so we can have as many implementations as we like.

I'm starting by adding a non-regex "Remove Prefixes" implementation in #233, which will accept multiple prefixes. The prefix that matches will be removed.

If in the future somebody really needs regex, we can add that as another implementation.

**Comment by chriskilding** [comment]
Implemented in #233
